File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/Kun8.java
package com.github.adamantcheese.chan.core.site.sites;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.adamantcheese.chan.core.model.Post;
import com.github.adamantcheese.chan.core.site.ChunkDownloaderSiteProperties;
import com.github.adamantcheese.chan.core.site.Site;
import com.github.adamantcheese.chan.core.site.SiteAuthentication;
import com.github.adamantcheese.chan.core.site.SiteIcon;
import com.github.adamantcheese.chan.core.site.common.CommonSite;
import com.github.adamantcheese.chan.core.site.common.MultipartHttpCall;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanActions;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanApi;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanCommentParser;
import com.github.adamantcheese.chan.core.site.common.vichan.VichanEndpoints;
import com.github.adamantcheese.chan.core.site.http.Reply;
import com.github.adamantcheese.chan.core.site.parser.CommentParserType;
import com.github.adamantcheese.common.ModularResult;
import com.github.adamantcheese.model.data.descriptor.ChanDescriptor;

import org.jetbrains.annotations.NotNull;

import java.util.Map;
import java.util.Objects;

import kotlin.Unit;
import okhttp3.HttpUrl;

public class Kun8
        extends CommonSite {
    private final ChunkDownloaderSiteProperties chunkDownloaderSiteProperties;
    public static final String SITE_NAME = "8kun";

    public static final CommonSiteUrlHandler URL_HANDLER = new CommonSiteUrlHandler() {
        private final String[] mediaHosts = new String[]{"media.8kun.top"};

        @Override
        public Class<? extends Site> getSiteClass() {
            return Kun8.class;
        }

        @Override
        public HttpUrl getUrl() {
            return HttpUrl.parse("https://8kun.top/");
        }

        @Override
        public String[] getNames() {
            return new String[]{"8kun"};
        }

        @Override
        public String[] getMediaHosts() {
            return mediaHosts;
        }

        @Override
        public String desktopUrl(ChanDescriptor chanDescriptor, @Nullable Long postNo) {
            if (chanDescriptor instanceof ChanDescriptor.CatalogDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .toString();
            } else if (chanDescriptor instanceof ChanDescriptor.ThreadDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .addPathSegment("res")
                        .addPathSegment(((ChanDescriptor.ThreadDescriptor) chanDescriptor).getThreadNo() + ".html")
                        .toString();
            } else {
                return getUrl().toString();
            }
        }
    };

    public Kun8() {
        chunkDownloaderSiteProperties = new ChunkDownloaderSiteProperties(true, true);
    }

    @Override
    public void setup() {
        setEnabled(true);
        setName(SITE_NAME);
        setIcon(SiteIcon.fromFavicon(getImageLoaderV2(), HttpUrl.parse("https://8kun.top/static/favicon.ico")));
        setBoardsType(BoardsType.INFINITE);
        setResolvable(URL_HANDLER);

        setConfig(new CommonConfig() {
            @Override
            public boolean siteFeature(SiteFeature siteFeature) {
                return super.siteFeature(siteFeature) || siteFeature == SiteFeature.POSTING
                        || siteFeature == SiteFeature.POST_DELETE;
            }
        });

        setEndpoints(new VichanEndpoints(this, "https://8kun.top", "https://sys.8kun.top") {

            @NonNull
            @Override
            public HttpUrl imageUrl(Post.Builder post, Map<String, String> arg) {
                HttpUrl url =
                        HttpUrl.parse("https://media.8kun.top/file_store/" + (arg.get("tim") + "." + arg.get("ext")));

                return Objects.requireNonNull(url, "image url is null");
            }

            @NonNull
            @Override
            public HttpUrl thumbnailUrl(Post.Builder post, boolean spoiler, int customSpoilters, Map<String, String> arg) {
                String ext;
                switch (arg.get("ext")) {
                    case "jpeg":
                    case "jpg":
                    case "png":
                    case "gif":
                        ext = arg.get("ext");
                        break;
                    default:
                        ext = "jpg";
                        break;
                }

                HttpUrl url =
                        HttpUrl.parse("https://media.8kun.top/file_store/thumb/" + (arg.get("tim") + "." + ext));

                return Objects.requireNonNull(url, "thumbnail url is null");
            }
        });

        setActions(new VichanActions(this, getOkHttpClient(), getSiteManager()) {
            @Override
            public ModularResult<Unit> setupPost(Reply reply, MultipartHttpCall call) {
                return super.setupPost(reply, call)
                        .mapValue(unit -> {
                            if (reply.chanDescriptor instanceof ChanDescriptor.ThreadDescriptor) {
                                // "thread" is already added in VichanActions.
                                call.parameter("post", "New Reply");
                            } else {
                                call.parameter("post", "New Thread");
                                call.parameter("page", "1");
                            }

                            return Unit.INSTANCE;
                        });
            }

            @Override
            public boolean requirePrepare() {
                // We don't need to check the antispam fields for 8chan.
                return false;
            }

            @Override
            public SiteAuthentication postAuthenticate() {
                return SiteAuthentication.fromUrl(
                        "https://sys.8kun.top/dnsbls_bypass.php",
                        "You failed the CAPTCHA",
                        "You may now go back and make your post"
                );
            }
        });

        setApi(new VichanApi(getSiteManager(), getBoardManager(), this));

        setParser(new VichanCommentParser(getMockReplyManager()));
    }

    @NotNull
    @Override
    public CommentParserType commentParserType() {
        return CommentParserType.VichanParser;
    }

    @NonNull
    @Override
    public ChunkDownloaderSiteProperties getChunkDownloaderSiteProperties() {
        return chunkDownloaderSiteProperties;
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/kun8/Kun8.kt
package com.github.k1rakishou.chan.core.site.sites.kun8

import com.github.k1rakishou.chan.core.site.SiteActions
import com.github.k1rakishou.chan.core.site.SiteAuthentication
import com.github.k1rakishou.chan.core.site.SiteConfiguration
import com.github.k1rakishou.chan.core.site.SiteEndpoints
import com.github.k1rakishou.chan.core.site.SiteUrlHandler
import com.github.k1rakishou.chan.core.site.common.CommonSite
import com.github.k1rakishou.chan.core.site.common.DefaultPostParser
import com.github.k1rakishou.chan.core.site.common.MultipartHttpCall
import com.github.k1rakishou.chan.core.site.common.vichan.VichanActions
import com.github.k1rakishou.chan.core.site.common.vichan.VichanApi
import com.github.k1rakishou.chan.core.site.common.vichan.VichanEndpoints
import com.github.k1rakishou.chan.core.site.parser.PostParser
import com.github.k1rakishou.chan.core.site.parser.SiteApi
import com.github.k1rakishou.common.ModularResult
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor.CatalogDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor.ThreadDescriptor
import com.github.k1rakishou.model.data.site.SiteBoards
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.Request

class Kun8 : CommonSite() {
  override val enabled: Boolean = true
  override val name: String = SITE_NAME
  override val commentParserType = SiteConfiguration.CommentParserType.VichanParser
  override val globalSearchType = SiteConfiguration.GlobalSearchType.SearchNotSupported
  override val siteIconUrl: HttpUrl = "https://media.128ducks.com/static/favicon.ico".toHttpUrl()
  override val boardsType = SiteConfiguration.BoardsType.Dynamic
  override val catalogType = SiteConfiguration.CatalogType.Static
  override val chunkedDownloaderConfig by lazy {
    SiteConfiguration.ChunkedDownloaderConfig(
      enabled = false,
      siteSendsCorrectFileSizeInBytes = false
    )
  }
  override val urlHandler: SiteUrlHandler by lazy { Kun8UrlHandler() }
  override val endpoints: SiteEndpoints by lazy { Kun8Endpoints(this) }
  override val api: SiteApi by lazy { VichanApi(siteManager, boardManager, this) }
  override val actions: SiteActions by lazy { Kun8Actions(this) }
  override val postParser: PostParser by lazy { DefaultPostParser(Kun8CommentParser(), archivesManager) }

  override fun hasSiteFeature(siteFeature: SiteConfiguration.SiteFeature): Boolean {
    return super.hasSiteFeature(siteFeature)
      || siteFeature === SiteConfiguration.SiteFeature.Posting
      || siteFeature === SiteConfiguration.SiteFeature.PostDeletion
  }

  private class Kun8Actions(
    kun8: Kun8
  ) : VichanActions(kun8, kun8.proxiedOkHttpClient, kun8.siteManager, kun8.replyManager) {
    override suspend fun boards(): Flow<SiteBoards> {
      val request = Request.Builder()
        .url(site.endpoints.boards().toString())
        .get()
        .build()

      val siteBoards = Kun8BoardsRequest(
        siteDescriptor = site.descriptor,
        boardManager = site.boardManager,
        request = request,
        proxiedOkHttpClient = site.proxiedOkHttpClient
      )
        .execute()
        .mapErrorToValue { error -> SiteBoards.Result.Error(error) }

      return flowOf(siteBoards)
    }

    override fun setupPost(
      replyChanDescriptor: ChanDescriptor,
      call: MultipartHttpCall
    ): ModularResult<Unit> {
      return super.setupPost(replyChanDescriptor, call)
        .mapValue {
          if (replyChanDescriptor is ThreadDescriptor) {
            // "thread" is already added in VichanActions.
            call.parameter("post", "New Reply")
          } else {
            call.parameter("post", "New Thread")
            call.parameter("page", "1")
          }
        }
    }

    override fun requirePrepare(): Boolean {
      // We don't need to check the antispam fields for 8chan.
      return false
    }

    override fun postAuthenticate(): SiteAuthentication {
      return SiteAuthentication.fromUrl(
        "https://sys.8kun.top/dnsbls_bypass.php",
        "You failed the CAPTCHA",
        "You may now go back and make your post"
      )
    }
  }

  private class Kun8Endpoints(kun8: Kun8) : VichanEndpoints(kun8, "https://8kun.top", "https://sys.8kun.top") {
    override fun imageUrl(boardDescriptor: BoardDescriptor, arg: Map<String, String>?): HttpUrl {
      requireNotNull(arg)

      val tim = requireNotNull(arg["tim"]) { "\"tim\" parameter not found" }
      val ext = requireNotNull(arg["ext"]) { "\"ext\" parameter not found" }
      val fpath = arg["fpath"]?.toIntOrNull() ?: 1

      val url = if (fpath == 1) {
        "https://media.128ducks.com/file_store/$tim.$ext".toHttpUrlOrNull()
      } else {
        "https://media.128ducks.com/${boardDescriptor.boardCode}/src/$tim.$ext".toHttpUrlOrNull()
      }

      return requireNotNull(url) { "image url is null" }
    }

    override fun thumbnailUrl(
      boardDescriptor: BoardDescriptor,
      spoiler: Boolean,
      customSpoilers: Int,
      arg: Map<String, String>?
    ): HttpUrl {
      requireNotNull(arg)

      if (spoiler) {
        return "https://media.128ducks.com/static/assets/${boardDescriptor.boardCode}/spoiler.png".toHttpUrl()
      }

      val tim = requireNotNull(arg["tim"]) { "\"tim\" parameter not found" }
      val fpath = arg["fpath"]?.toIntOrNull() ?: 1

      val extension = when (val ext = requireNotNull(arg["ext"]) { "\"ext\" parameter not found" }) {
        "jpeg", "jpg", "png", "gif" -> ext
        else -> "jpg"
      }

      val url = if (fpath == 1) {
        "https://media.128ducks.com/file_store/thumb/$tim.$extension".toHttpUrlOrNull()
      } else {
        // Oldstyle images seems to always have "jpg" extension. But even if some of them don't
        // (I couldn't find any but there might be some) there is no way to figure out the true
        // extension because API only sends the original image extension.
        "https://media.128ducks.com/${boardDescriptor.boardCode}/thumb/$tim.$extension".toHttpUrlOrNull()
      }

      return requireNotNull(url) { "thumbnail url is null" }
    }

    override fun boards(): HttpUrl {
      return root.builder().s("boards.json").url()
    }
  }

  private class Kun8UrlHandler : CommonSiteUrlHandler() {
    override val mediaHosts = arrayOf(
      "https://media.8kun.top/".toHttpUrl(),
      "https://media.128ducks.com/".toHttpUrl()
    )

    override val url: HttpUrl = "https://8kun.top/".toHttpUrl()

    override fun desktopUrl(chanDescriptor: ChanDescriptor, postNo: Long?, postSubNo: Long?): String? {
      return when (chanDescriptor) {
        is CatalogDescriptor -> {
          url.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .toString()
        }
        is ThreadDescriptor -> {
          url.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .addPathSegment("res")
            .addPathSegment(chanDescriptor.threadNo.toString() + ".html")
            .toString()
        }
        else -> null
      }
    }
  }

  companion object {
    const val SITE_NAME = "8kun"
  }

}
